### PR TITLE
Fix JsonSyntaxException when creating API command response #4355 

### PR DESCRIPTION
**Pull request by GabrielBrascher**
## Description

Added missing `\"` (from `"*****"` to `"*****\""`) at _obfuscatePassword method from AsyncJobManagerImpl.java#L485_
https://github.com/apache/cloudstack/blob/0f3f2a09370a18301db28ec3d28efe746b6437c9/framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/impl/AsyncJobManagerImpl.java#L485

The lack of `"` caused a JsonSyntaxException exception when the host response had a password (e.g. if the host has out of band enabled e.g. IPMI or Redfish).

Fixes: #4355

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?

- JUnit test
- manual tests

### CODE REVIEW

**Comment by GabrielBrascher** [comment]
@blueorangutan package

**Review by DaanHoogland — APPROVED** [review]
lgtm

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
did work @GabrielBrascher 
```
======================================================================
ERROR: Tests degraded HA state when agent is stopped/killed
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/marvin/tests/smoke/test_hostha_kvm.py", line 355, in test_hostha_kvm_host_degraded
    self.assertIssueCommandState('ON', 'On')
  File "/marvin/tests/smoke/test_hostha_kvm.py", line 493, in assertIssueCommandState
    self.issuePowerActionCmd(command)
  File "/marvin/tests/smoke/test_hostha_kvm.py", line 539, in issuePowerActionCmd
    raise e
Exception: Job failed: {jobprocstatus : 0, created : u'2020-10-08T02:10:21+0000', completed : u'2020-10-08T02:10:21+0000', cmd : u'org.apache.cloudstack.api.command.admin.outofbandmanagement.IssueOutOfBandManagementPowerActionCmd', userid : u'a8138930-08c3-11eb-8574-1e00ca0138c9', jobstatus : 2, jobid : u'f8439839-f821-4bd6-b7a3-820345992038', jobresultcode : 530, jobresulttype : u'object', jobresult : {errorcode : 530, errortext : u'com.google.gson.stream.MalformedJsonException: Unterminated object near "password":"p*****"","action":"ON","desc'}, accountid : u'a812be97-08c3-11eb-8574-1e00ca0138c9'}
```
note that in this case the replacement yield a double double quote, which is one in excess.

**Comment by GabrielBrascher** [comment]
@DaanHoogland what about removing the password from the JSON response?

**Comment by DaanHoogland** [comment]
@GabrielBrascher If sometimes an extra '\"' is added and sometimes not, this is still going to be an issue if you try to remove it. I'd rather replace the exact "something" with "****" or "s***". it seems that we are brushing a bug under the carpet otherwise.

File: framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/impl/AsyncJobManagerImpl.java [hunk 1]
```diff
@@ -475,14 +475,14 @@ public AsyncJob queryJob(final long jobId, final boolean updatePollTime) {
...
-    private String obfuscatePassword(String result, boolean hidePassword) {
+    public String obfuscatePassword(String result, boolean hidePassword) {
         if (hidePassword) {
             String pattern = "\"password\":";
             if (result != null) {
                 if (result.contains(pattern)) {
                     String[] resp = result.split(pattern);
                     String psswd = resp[1].toString().split(",")[0];
-                    result = resp[0] + pattern + psswd.replace(psswd.substring(2, psswd.length() - 1), "*****") + "," + resp[1].split(",", 2)[1];
+                    result = resp[0] + pattern + psswd.replace(psswd.substring(2, psswd.length() - 1), "*****\"") + "," + resp[1].split(",", 2)[1];
```
Comment (rhtyd) [inline framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/impl/AsyncJobManagerImpl.java:485]: Why not complete replace with the `****`?

**Comment by GabrielBrascher** [comment]
@DaanHoogland I proposed removing the password because I don't see much of use in having it on the response, it seems that does not aggregate relevant information.

**Comment by DaanHoogland** [comment]
It is adding value but only if having a password is optional, @GabrielBrascher. And i think hiding it should be optional as well as a private cloud operator might opt for having it visible. I am not going to argue that anymore, btw.

File: framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/impl/AsyncJobManagerImpl.java (hunk 1 above)
Comment (GabrielBrascher) [reply framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/impl/AsyncJobManagerImpl.java:485]: I am not sure if replacing the password with `****` would then affect the intention of the password response. 
I think that the idea was to offer admins a hint on the password.

But that is good to keep in mind, thanks for the heads up @rhtyd. What do you think @DaanHoogland?

**Comment by GabrielBrascher** [comment]
@DaanHoogland I found out why some cases the regex failed and some not. Updated code to fix it (for good, I hope). Also added test case addressing both situations.

The regex uses comma to set the split separator: `String psswd = resp[1].toString().split(",")[0];`
This works on cases such as:
```
...,"password":"password","action":"OFF"...
```

However, the specific case that I got with OOBM it has the following JSON String:

```
"username":"root","password":"password"},"resourcestate":"PrepareForMaintenance","hahost":false
```

With that said, the code has been updated to deal with both cases: `"password":"password"},` and `"password":"password",`.

**Comment by GabrielBrascher** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@rhtyd no further comments?

**Review by rhtyd — APPROVED** [review]

**Pull request merged**